### PR TITLE
Test fixes for external C++/Redis tests

### PR DESCRIPTION
**Pull request by Porges**
The way that `exec_core` was written, the C++ & Redis tests couldn’t actually fail even if they returned failure codes. It handles a `Result<Result<_,_>>` but only one layer of `Result` was unwrapped. If the inner `Result` was a test failure the test would still succeed.

Also fixup and improve the C++ code so it checks all results, add `-Wall` flag to build step. While doing this I noticed that the C bindings return an expected error code upon success…

### CODE REVIEW

**Comment by radu-matei** [comment]
Would you mind signing your commits, please, @Porges ?
https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-new-gpg-key-to-your-github-account

File: tests/integration.rs [hunk 1]
```diff
@@ -260,12 +260,15 @@ mod runtime {
...
-    fn exec_core<T>(mut store: Store<Context<T>>, mut instance: Instance) -> Result<()> {
-        let t = test::Test::new(&mut store, &mut instance, |host| {
+    fn exec_core<T>(mut store: Store<Context<T>>, instance: Instance) -> Result<()> {
+        let t = test::Test::new(&mut store, &instance, |host| {
             host.test_data.as_mut().unwrap()
         })?;
-        let _ = t.test(&mut store).expect("Error running the test method");
-        Ok(())
+        let result = t.test(&mut store).expect("Error running the test method");
+        match result {
```
Comment (radu-matei) [inline tests/integration.rs:268]: Well, that is embarrassing.. :)
Thanks for catching this!

File: tests/modules/cache-cpp/lib.cpp [hunk 2]
```diff
@@ -7,27 +7,62 @@
 #include "bindings/wasi-cache.h"
 #include "bindings/test.h"
 
+// TODO: for some reason bindings are returning
+// -1 upon success instead of WASI_CACHE_ERROR_SUCCESS (0)
```
Comment (radu-matei) [inline tests/modules/cache-cpp/lib.cpp:11]: Huh, this is weird.
The errors seem to be properly defined in the bindings:
```cpp
  typedef uint8_t wasi_cache_error_t;
  #define WASI_CACHE_ERROR_SUCCESS 0
  #define WASI_CACHE_ERROR_ERROR 1
```

File: tests/modules/cache-cpp/lib.cpp [hunk 3]
```diff
@@ -7,27 +7,62 @@
...
+    result = wasi_cache_delete(&skey);
+    if (result != SUCCESS_VALUE) {
+        fprintf(stderr, "Failed to delete value\n");
+        return TEST_ERROR_FAILURE;
+    }
 
-    assert(svalue->len == ret->len);
+    free(svalue.ptr);
+    free(skey.ptr);
+    free(ret.ptr);
```
Comment (radu-matei) [inline tests/modules/cache-cpp/lib.cpp:64]: Well, there is a reason I mostly write Rust..

File: tests/modules/cache-cpp/lib.cpp (hunk 2 above)
Comment (Porges) [reply tests/modules/cache-cpp/lib.cpp:11]: Right, I don’t know what’s happening here, it’s probably something in `wit-bindgen`… but every function ends with a line like: 

```c
return variant2.tag ? variant2.val.err : -1;
```

**Review by radu-matei — APPROVED** [review]
Thanks a lot for catching this.

I'm a bit surprised at the error value, feel free to open that as an issue so we can follow up on it.

Once the commit(s) are signed, this is LGTM.

Thanks a lot for your contribution!

**Comment by Porges** [comment]
> Would you mind signing your commits, please, @Porges ? https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-new-gpg-key-to-your-github-account

Signed now, hadn’t gotten it all set up in Codespaces before 😊

File: tests/modules/cache-cpp/lib.cpp (hunk 2 above)
Comment (radu-matei) [reply tests/modules/cache-cpp/lib.cpp:11]: Ok, I'll look into that at some point.
Thanks for the heads up!

Comment (Porges) [reply tests/modules/cache-cpp/lib.cpp:11]: For some reason bindgen emits a hardcoded value here:
https://github.com/bytecodealliance/wit-bindgen/blob/e9c7c0a3405845cecd3fe06f3c20ab413302fc73/crates/gen-c/src/lib.rs#L1863

Comment (Porges) [reply tests/modules/cache-cpp/lib.cpp:11]: @radu-matei figured out what is going on here, the nesting of errors was confusing me before. `WASI_CACHE_ERROR_SUCCESS` is only returned in the `Err(test::Error::Success)` case, it is explicitly _not_ the `Ok` case! On the other hand, the `Ok` case should probably have its own constant to compare to, which `wit-bindgen` does not currently create.

**Comment by Porges** [comment]
> Thanks a lot for catching this.
> 
> I'm a bit surprised at the error value, feel free to open that as an issue so we can follow up on it.
> 
> Once the commit(s) are signed, this is LGTM.
> 
> Thanks a lot for your contribution!

@radu-matei figured out what is going on here with the error value. The nesting of errors was confusing me before. `WASI_CACHE_ERROR_SUCCESS` is only returned in the `Err(test::Error::Success)` case, it is explicitly _not_ the “success” case of the `unexpected`! On the other hand, the `Ok` case should probably have its own constant to compare to, which `wit-bindgen` does not currently create.

**Pull request merged**